File: benchmarks/random_episodes.cpp
#include <iostream>
#include <random>

#include "buchberger.h"


int main(int argc, char* argv[]) {
  if (argc != 3) {
    std::cout << "Usage: random_episodes <dist> <episodes>" << std::endl;
    return 1;
  }

  LeadMonomialsEnv env{argv[1], false, true, 2};
  std::default_random_engine rng;

  double total_reward = 0.0;
  int episodes = std::stoi(argv[2]);

  for (int i = 0; i < episodes; i++) {
    env.reset();
    while (!env.state.empty()) {
      std::uniform_int_distribution<int> dist{0, env.state.size() / env.cols - 1};
      int action = dist(rng);
      double reward = env.step(action);
      total_reward += reward;
    }
  }

  std::cout << total_reward / episodes << std::endl;
}

File: scripts/random_episodes.py
#!/usr/bin/env python
"""Benchmark running random episodes on the environment."""

import numpy as np
import sys

from deepgroebner.buchberger import LeadMonomialsEnv
from deepgroebner.wrapped import CLeadMonomialsEnv


def run_episode(agent, env):
    """Return total reward from one episode of agent in env."""
    state = env.reset()
    done = False
    total_reward = 0
    while not done:
        action = agent.act(state)
        state, reward, done, _ = env.step(action)
        total_reward += reward
    return total_reward


class RandomAgent:
    """Agent which selects randomly in LeadMonomialsWrapper environment."""

    def act(self, state):
        return np.random.choice(len(state))


if __name__ == "__main__":
    agent = RandomAgent()
    if len(sys.argv) == 4 and sys.argv[-1] == '--python':
        env = LeadMonomialsEnv(sys.argv[1], k=2)
        returns = [run_episode(agent, env) for _ in range(int(sys.argv[2]))]
        print(np.mean(returns))
    elif len(sys.argv) == 3:
        env = CLeadMonomialsEnv(sys.argv[1], k=2)
        returns = [run_episode(agent, env) for _ in range(int(sys.argv[2]))]
        print(np.mean(returns))
    else:
        print("Usage: random_episodes.py <dist> <episodes>")
